### PR TITLE
Add take_or_recover() which replaces the value with a recovery value on panics

### PR DESCRIPTION
**Pull request by sdroege**
This seems like a good idea in cases where panics are caught at a safe point in the application and you don't want your whole application to disappear because of that.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -53,9 +54,90 @@ fn it_works() {
...
+///     // Do more stuff
+///     Foo // Return new Foo from closure, which goes back into the &mut Foo
+/// });
+/// ```
+pub fn take_or_recover<T, F>(mut_ref: &mut T, recover: T, closure: F)
+  where F: FnOnce(T) -> T {
+    use std::ptr;
+    unsafe {
+        let old_t = ptr::read(mut_ref);
+        let new_t = panic::catch_unwind(panic::AssertUnwindSafe(|| closure(old_t)));
```
Comment (sdroege) [inline src/lib.rs]: Using AssertUnwindSafe here for the whole closure and old_t should be safe as we just intercept the panic, write our pointer (to the mutable reference that can't be modified by the closure in safe code!) and then pass on the panic.

Catching the panic in a recovering way without passing it onwards and the safeness involved with that would be the job of the caller if it wishes so, just like in cases without usage of take_or_recover()

**Comment by Sgeo** [comment]
Sorry I haven't looked at this, I'm not very good at maintaining interest in stuff.

What's the difference between this and using `std::mem::replace` twice?

**Comment by sdroege** [comment]
You mean doing

```Rust
let old_t = mem::replace(mut_ref, recover);
let new_t = closure(old_t);
mem::replace(mut_ref, new_t);
```

That should indeed be equivalent and much simpler. If you'd like to merge a change like that, I can update the pull request.

What could also make sense, is to use a `FnOnce() -> T` instead of `T` for `recover`. That way it would be possible to prevent creation of the "recovery value" unless things go wrong, which brings performance advantages in the "good" and more common case. However it would require doing the manual panic catching again. IMHO that variant is nicer, and I could change the pull request accordingly if you also agree.

**Comment by Sgeo** [comment]
Functions for both `FnOnce() -> T` and `T` for `recover` might make sense, to avoid people using the closure version out of convenience if not necessary, but better to use the safe implementation.

Also, I'm still a bit confused about `std::panic`, and I saw an implementation in the RFC that did not rely on that. Could either the entire codebase be switched to `std::panic` or written to not use it?

**Comment by sdroege** [comment]
> Functions for both FnOnce() -> T and T for recover might make sense, to avoid people using the closure version out of convenience if not necessary, but better to use the safe implementation.

What do you mean with that? The non-closure version would be more convenient, and usually slower. I'd prefer to only have a variant using a closure

> Also, I'm still a bit confused about std::panic, and I saw an implementation in the RFC that did not rely on that. Could either the entire codebase be switched to std::panic or written to not use it?

I think `std::panic` is more explicit and would prefer to use that. Your opinion?
Not sure this can be written without unsafe code and without `std::panic`.

**Comment by Sgeo** [comment]
> What do you mean with that? The non-closure version would be more convenient, and usually slower. I'd prefer to only have a variant using a closure

I guess my thought process was that I'd prefer people to avoid using the closure version if the non-closure version was usable, but that's only because the latter wouldn't need unsafe code in implementation. It's probably not a real concern, come to think of it, and polluting the namespace wouldn't be ideal.

> I think std::panic is more explicit and would prefer to use that. Your opinion?
Not sure this can be written without unsafe code and without std::panic.

It's mostly that I don't totally understand `AssertUnwindSafe` to understand when that's safe to use. I do know at one point someone did suggest to use `std::panic` when it became stable.

Once I understand it, I'm fine with using `std::panic`, but think in that case maybe `take` should be switched over too? What do you think?

Non-`std::panic` version of your function: https://github.com/rust-lang/rfcs/pull/1736#issuecomment-261350311

**Comment by Sgeo** [comment]
Also need to be careful with the closure version, to handle the situation where the recovery closure itself panics...

**Comment by sdroege** [comment]
Ok, I'll update the pull request on Monday accordingly. Thanks for your comments :)

**Comment by sdroege** [comment]
Updated accordingly

**Comment by sdroege** [comment]
Thanks! Are you planning a release some time soon?

**Comment by Sgeo** [comment]
Working on new functionality over at https://github.com/Sgeo/take_mut/tree/scoped

Once it's done and I ask people to review for safety (maybe on Reddit) I'll merge and release.

**Pull request merged**